---
title: "Lecture 17: Bulk RNA-seq analysis"
output:
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true

---
```{r}
knitr::opts_chunk$set(fig.width=5, fig.height=5) 
```


```{r, include=FALSE}
# All necessary importat
library(rtracklayer)
library(Rsamtools)
library(GenomicAlignments)
library(DESeq2)
library(EnhancedVolcano)
library(goseq)
library(pheatmap)
library(RColorBrewer)
```


# Count matrices

STAR and other aligners generate bam files with information about the reads mapping to genomic co-ordinates. Count matrix generation is the process of using this data to generate a matrix of gene expression counts for each sample. In addition to the bam files, we also need transcript information which is available through gtf files. GTF files for all organisms can be downloaded using [ENSEMBL](https://www.ensembl.org/index.html?redirect=no). <b> Important </b>: Make sure that the genome index and gtf file are from the same version of the genome. Current version of the human genome is [GRCH38/hg38](https://www.ensembl.org/Homo_sapiens/Info/Index) and the mouse genome is [GRCm39/mm10](https://www.ensembl.org/Mus_musculus/Info/Index). 

## Data - BAM files

We will use a small dataset to practice count matrix generation. The data is from the `Pasilla` [knockdown experiment](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3032923/). We have bam files from two untreated samples available [here](https://drive.google.com/file/d/1VWvwatiE4WbMpocCm3W1rdCtakkqdAI6/view?usp=sharing). The bam files were subset to include only `chr4`. Copy the bam files over to the `data` directory 

```{r}
dir('data/', pattern='bam')
```

## GTF file
Lets download the GTF file. The bams were generated using `Release 62` of the Drosophila genome.

```{r}
# This URL can be copied over from ENSEMBL
fl <- paste0("ftp://ftp.ensembl.org/pub/release-62/gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP5.25.62.gtf.gz")

# Create a destination file name 
gffFile <- file.path('data/', basename(fl))
print(gffFile)

# Download the gtf file to the data directory
download.file(fl, gffFile)

```

```{r}
# Verify that the file is downloaded 
dir('data/')
```

Next step is to load the gtf file into R. We will use the `rtracklayer` package for importing the gtf file

```{r}
gff0 <- import(gffFile)
```
Examine the `gtf` file. As you will see, the imported gtf is a `GRanges` object with all the relevant information about transcript definition. 
```{r}
gff0
```

### GRanges refresher exercise
```{r}
# What are the different source of genes in the gff?
unique(gff0$source)

# What are the different entry `type`s
unique(gff0$type)
```

Since our bam files contain data from `chr4` only, subset the gff to contain genes from only chromosome 4. Note that this is not strictly necessary, but will only speed up the processing. This will also give you an idea of how to subset different gene sets of interest if you do not need to generate a full count matrix. 

```{r}
# Exons from protein coding genes in chromosome 4
idx <- mcols(gff0)$source == "protein_coding" &
            mcols(gff0)$type == "exon" &
            seqnames(gff0) == "4"
gff <- gff0[idx]
# Adjust seqnames to match Bam files. This is important because bam files have seqnames as "chr1", "chr2", etc whereas ENSEMBL gtfs have seqnames as "1", "2". 
seqlevels(gff) <- paste("chr", seqlevels(gff), sep="")
```

Verify that your filtering worked:
```{r}
unique(gff$source)
unique(gff$type)
unique(seqnames(gff))
```

Final step: We need to group the codons by gene id / gene name / transcript id
```{r}
chr4genes <- split(gff, mcols(gff)$gene_id)

# Length will equal to number of genes
length(chr4genes)

# Names of genes
names(chr4genes)
```


<b> Quick exercise </b></br>
```{r}
# Extract coding sequences from protein coding genes on chromosome 5
```


```{r}
# Extract pseudodgenes and tRNAs from all chromosomes
```

## Let's count!

```{r}
# Define bam parameters using `ScanBamParam`
param <- ScanBamParam(
              what='qual',
              which=GRanges("chr4", IRanges(1, 1e6)),
              flag=scanBamFlag(isUnmappedQuery=FALSE, isPaired=NA),
 tag="NH")

# Bam files 
bamFiles <- c("data/untreated1_chr4.bam", "data/untreated3_chr4.bam")
```

`summarizeOverlaps` from `GenomicAlignments` package is the function to use for creating count matrices

```{r}
genehits <- summarizeOverlaps(chr4genes, bamFiles, mode="Union")
genehits
```




`geneHits` is an object of `RangedSummarizedExperiment`. `SummarizedExperiment` is a class for holding counts and other matrices along with metadata information about the samples, genes and the experiment itself. An illustration of `SummarizedExperiment` is shown below. More details [here](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html). 

![](assets/SummarizedExperiment.jpg)

`RangedSummarizedExperiment` is same as `SummarizedExperiment` such that the row data is a `GRanges` object. In this case features are genes. Examine the counts, sample information and gene information 

```{r}
head(assays(genehits)[['counts']])
```

```{r}
head(rownames(genehits))
```

```{r}
colnames(genehits)
```

`SummarizedExperiment` can be subset by rows or columns

<p>
<b>Important</b> A key parameter in `summarizeOverlaps` is `mode`. `Union` is a general safe bet.
![](assets/summarizeoverlaps.jpg)
</p>




# Differential analysis

Differential analysis is used to identify differentially expressed genes in different conditions. We will use the [TCGA Glioblastoma](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga/studied-cancers/glioblastoma) dataset for differential analysis.  The dataset contains a total of 114 patient samples and is annotated as the following subtypes: Proneural, Neural, Mesenchymal and G-CIMP. The counts data is available [here](https://drive.google.com/file/d/1aaV8QFcRPXF3oOzGdwKpqmRWyd4DrMKN/view?usp=sharing) and the patient metadata (subtypes) is available [here](https://drive.google.com/file/d/1NcGZcyhpJ1QxUGL4aYpe7KsPLUXfQ9Kd/view?usp=sharing).

## Load the data 
```{r}
# These are simple csv files, which we will load using read.csv 
# Make sure to set check.names to FALSE - R annoyingly resets special characters to dots
counts <- read.csv('data/tcga_gbm_clean_counts.csv', check.names=FALSE, row.names=1)
metadata <- read.csv('data/tcga_gbm_clean_patient_metadata.csv', check.names=FALSE, row.names=1)
```

Examine the data - we will just look at a few things to make sure all looks good
```{r}
# Dimensionality of the counts and metadata should match
dim(counts)
dim(metadata)

# Sample names should be same 
sum(colnames(counts) != rownames(metadata))

head(metadata)
```

There are four different subtypes of GBM samples in this dataset: Mesenchymal, Neural, G-CIMP, Proneural
```{r}
unique(metadata[, 'Subtype'])
```

## SummarizedExperiment

We will first create a `SummarizedExperiment` object which will contain all the information we need

```{r}
gbmSE <- SummarizedExperiment(assays = list(counts = as.matrix(counts)), colData = metadata)
gbmSE
```

Subset `gbmSE` to include only Proneural and Mesenchymal subtypes and call the new object ```gbmSE_sub```. 
```{r}
gbmSE_sub = gbmSE[, colData(gbmSE)[, 'Subtype'] %in% c('Proneural', 'Mesenchymal')]
gbmSE_sub

```

<b> Quick exercise </b></br>
```{r}
# Create 'gbmSE_mes_neur' with Mesenchymal and Neural subtypes
```

## Differential analysis - simple

We will use `DESeq2` for differential analysis. Details about DESeq2 are available [here](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8). [edgeR](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/) is another popular tool for differential analysis for bulk counts data. Practically, the differences are minimal and they generally both work well. A detailed tutorial of theory and extensive use-cases for DESeq2 is found [here](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). Lecture is based on almost entirely on the linked vignette.

As a start, lets identify differentially expressed genes between proneural and mesenchymal subtypes. First we need to set up a `DESeqDataSet`. The `design` parameter includes all the sample metadata that need to be used to identify differential expression (More on this later). Here, we only have one metadata ie. Subtype. This will go to the `design` parameter of `DESeqDataSet` function.

```{r}
dds <- DESeqDataSet(gbmSE_sub, design = ~ Subtype)
dds

# Reset the levels to use proneural as reference 
dds$Subtype <- relevel(dds$Subtype, ref = "Proneural")
```

Perform differential analysis. 
```{r}
dds <- DESeq(dds)
res <- results(dds)
mes_vs_pro_pairwise = res
```

<b>Done!</b>

Since we used proneural subtype as reference, genes with negative fold change are higher in proneural and genes with positive fold changes are higher in mesenchymal. Examine the top differentially expressed genes in each condition

```{r}
# Sort by adjusted p-values 
resOrdered <- res[order(res$padj),]

# Top 10 proneural genes
head(resOrdered[resOrdered[, 'log2FoldChange'] < 0,  ])


# Top 10 mesenchymal genes
head(resOrdered[resOrdered[, 'log2FoldChange'] > 0,  ])

```

MA-plots are a great way to visualize differential results. The plots are simple: X-axis represent mean across both conditions and y-axis represents the log fold change. 

```{r}
plotMA(res, ylim=c(-2.5, 2.5),)
```



Another popular visualization is called the `volcano` plot which compares the fold change with p-values. 

```{r}
  EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue')
```

<b> Quick Exercise </b></br>
What can you say about differentially expressed genes in proneural and mesenchymal subtypes?



## A view behind the blackbox


![](assets/deseq2.jpg)

<b>Key points:</b></br>

* Each gene is modeled as a function of the specified conditions (See example below)
* Rather than fit a linear model on log counts, DESeq2 uses negative binomial distribution which models the underlying counts better. 
* In the model, greater slope implies greater difference 
* The dispersion estimates, representation of variance, is computed for each gene separately and information is shared across genes to derive a final estimate.

![](assets/lm.jpg)


<b> A fun exercise for later </b> </br>
Perform linear model fits for each gene using normalized log-transformed counts. Compare the slopes to the fold change estimates from DESeq2.



## Differential analysis - multiple conditions

Let us now consider the full dataset with all four conditions. Our goal is to identify differntially expressed genes in each condition relative to samples of all other conditions.

```{r}
# Set up the DESeq data object
dds <- DESeqDataSet(gbmSE, design = ~ Subtype + 0)
dds <- DESeq(dds)
```

`contrasts` can be used to specify the comparisons of interest. Lets compare results using mesenchymal -vs- proneural comparison

```{r}
mes_vs_pro = results(dds, contrast=c("Subtype", "Mesenchymal", "Proneural"))
mes_vs_pro
```

Compare to previous resuls 
```{r}
plot(res[, 'log2FoldChange'], 
     mes_vs_pro[rownames(res), 'log2FoldChange'], 
     xlab='Stand alone fold change', ylab='Full dataset comparison', 
     main='Mesenchymal v Proneural')
```

The `contrasts` can be vectors too!
```{r}
# Key: Pay attention to the listValues parameter
mes_vs_rest = results(dds, contrast=list(
                                      c("SubtypeMesenchymal"), 
                                      c("SubtypeProneural", "SubtypeNeural", 'SubtypeG.CIMP')), 
                      listValues=c(1, -1/3))
mes_vs_rest
```

Compare to mesenchymal -vs- proneural resuls 
```{r}
plot(mes_vs_rest[, 'log2FoldChange'], 
     mes_vs_pro[rownames(res), 'log2FoldChange'], 
     xlab='Mes -vs- Rest', ylab='Mes -vs- Pro', 
     )
```

```{r}
  EnhancedVolcano(mes_vs_rest,
    lab = rownames(mes_vs_rest),
    x = 'log2FoldChange',
    y = 'pvalue')
```

<b> Quick Exercise </b> </br>
Perform analysis of proneural -vs- rest and compare to proneural -vs- mesenchymal results.


## Differential analysis - technical differences
Often times, data are generated in multiple batches or different settings (E.g: single-end -vs- paired-end reads). These technical differences can lead to artifical differences between biological conditions. Such technical variables can be accounted for using DESeq by specifying them when constructing the `DESeqDataSet` object. 

```{r, eval=FALSE}
dds <- DESeqDataSet(gbmSE, design = ~ batch + Subtype + 0)
```


## Gene Ontology analysis

Gene ontology or pathway enrichments can be used to understand the collective behaviour of differentially expressed genes. The goal of this analysis is to check if the differentially expressed genes are enriched in certain known pathways, cell types, or biological processes.

![](assets/hypergeom.jpg)

There are a number of interesting gene sets available for exploration. [MSigDB](http://www.gsea-msigdb.org/gsea/msigdb) by the Broad Institute is an excellent source of these databases. MSigDB also has a simple interface to test for enrichments among differentially expressed genes.

Gene ontology analysis can be performed using the `goseq` package. 

First, extract the genes that are significantly high in proneural -vs- mesenchymal comparison with a `fold change > 2` and adjusted `p-value < 1e-2`

```{r}
# Binary vector indicating which gene is significant
proneural_genes = ((mes_vs_pro_pairwise[, 'log2FoldChange'] < -2) & (mes_vs_pro_pairwise[, 'padj'] < 1e-2))
names(proneural_genes) <- rownames(mes_vs_pro_pairwise)

# Setup background for goseq
pwf <- nullp(proneural_genes, 'hg19', 'geneSymbol')
enriched_pathways <- goseq(pwf, 'hg19', 'geneSymbol', test.cats='GO:BP')
head(enriched_pathways)
```

<b> Quick exercise </b></br>
```{r}
# Repeat the above analysis for mesenchymal genes in the mes -vs- pro pairwise comparison
```



# Visualization

Differential analysis was performed on the raw counts. For visualization and measuring distances between samples, we will need to account for the following factors

* Total reads or fragments between samples can be very different and therefore, the data needs to be normalized
* Variance is a function of the mean in count data types and therefore needs to be stabilized. 

`Variance stabilizing transformation (vst)` is a very effective way to achieve these. More details can be found [here](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations). `vst` can be performed using:

```{r}
vsd <- vst(dds, blind=FALSE)
```

## Heatmap of gene expression

Heatmaps are a very powerful way to visualize the data and understand the relationships between samples. We will extract the top 100 genes for each subtype compared to the rest of the subtype and plot the expression of these genes on a heatmap.


First, extract the top 100 genes for each condition
```{r}
n_genes <- 100
genes <- c()
subtypes <- c('SubtypeProneural', 'SubtypeMesenchymal', 
                'SubtypeNeural', 'SubtypeG.CIMP')
for (subtype in subtypes){
    subtype_res = results(dds, contrast=list(
                                      c(subtype), 
                                      setdiff(subtypes, c(subtype))), 
                      listValues=c(1, -1/3))
    
    resOrdered <- subtype_res[order(subtype_res$padj),]
    subtype_genes <- rownames(resOrdered[resOrdered[, 'log2FoldChange'] > 0,  ])[1:n_genes]
    genes <- c(genes, subtype_genes)

}
```


Second, create a matrix containing the above set of genes and all samples. It is highly recommended to z-score (subtract the mean, divide by standard deviation across samples) the expression values for highlighting the contrasts
```{r}
norm_matrix <- assay(vsd[genes, ])
scaled_exprs <- t(scale(t(norm_matrix)))
scaled_exprs[scaled_exprs > 2] = 2
scaled_exprs[scaled_exprs < -2] = -2
```


Finally, plot the heatmap!
```{r}
df <- as.data.frame(colData(dds)[,"Subtype"])
colnames(df) = 'Subtype'
rownames(df) <- colnames(dds)
pheatmap(scaled_exprs, cluster_rows=TRUE, 
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=TRUE, annotation_col=df)
```

What are your observations?


## Heatmap of sample to sample distances

Lets use the subset of genes identified to find distances between samples and plot them as a heatmap.

```{r}
# Compute distances 
sampleDists <- dist(t(norm_matrix))
sampleDistMatrix <- as.matrix(sampleDists)
sampleDistMatrix[sampleDistMatrix > 50] = 50

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,
         show_rownames=FALSE, show_colnames=FALSE,
         annotation_col=df, annotation_row=df)
```


<b>Quick exercise</b> </br>
Generate heatmap using all genes


## PCA
Visualizing using PCA can be very effective when dealing with a large number of samples. The following function plots the first two components.

```{r}
plotPCA(vsd, intgroup=c("Subtype"))

```

<b> A fun exercise for later </b> </br>

 * Train a classifier between a pair of subtypes or all cell types. 
 * Identify the genes that are most informative for the classification 
 * Compare the results to differential analysis between the same pair of subtypes 
 * What do you think are the advantages and disadvantages of using classification instead of differential analysis. 

